lib/HLSL/HLModule.cpp

```diff
@@ -1028,51 +1033,52 @@ static Value *MergeGEP(GEPOperator *SrcGEP, GetElementPtrInst *GEP) {
     Indices.append(SrcGEP->op_begin() + 1, SrcGEP->op_end());
     Indices.append(GEP->idx_begin() + 1, GEP->idx_end());
   }
-  if (!Indices.empty())
-    return Builder.CreateInBoundsGEP(SrcGEP->getSourceElementType(),
-                                     SrcGEP->getOperand(0), Indices,
-                                     GEP->getName());
-  else
-    llvm_unreachable("must merge");
+
+  DXASSERT(!Indices.empty(), "must merge");
+  Value *newGEP = Builder.CreateInBoundsGEP(nullptr, SrcGEP->getOperand(0), Indices, Name);
+  GEP->replaceAllUsesWith(newGEP);
+  if (Instruction *I = dyn_cast<Instruction>(GEP))
+    I->eraseFromParent();
+  return newGEP;
 }
 
 void HLModule::MergeGepUse(Value *V) {
-  for (auto U = V->user_begin(); U != V->user_end();) {
-    auto Use = U++;
-
-    if (GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(*Use)) {
-      if (GEPOperator *prevGEP = dyn_cast<GEPOperator>(V)) {
-        // merge the 2 GEPs
-        Value *newGEP = MergeGEP(prevGEP, GEP);
-        // Don't need to replace when GEP is updated in place
-        if (newGEP != GEP) {
-          GEP->replaceAllUsesWith(newGEP);
-          GEP->eraseFromParent();
-        }
-        MergeGepUse(newGEP);
+  SmallVector<Value*, 16> worklist;
+  auto addUsersToWorklist = [&worklist](Value *V) {
+    if (!V->user_empty()) {
+      // Add users in reverse to the worklist, so they are processed in order
+      // This makes it equivalent to recursive traversal
+      size_t start = worklist.size();
+      worklist.append(V->user_begin(), V->user_end());
+      size_t end = worklist.size();
+      std::reverse(worklist.data() + start, worklist.data() + end);
```

Comment:
This is the odd looking bit, but is necessary to preserve the iteration order in later code that dynamically adds resources when embedded in structures.  The order those resources are dynamically added impacts the register allocation, even though it should only be based on the order of declaration.  This is a bug that needs to be fixed separately, but this careful iteration order preserves the prior behavior at least, rather than reversing some orderings.